{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import argparse\n",
    "import sys\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "dir_data_raw = './data/raw/' # raw data folder\n",
    "dir_data_out = './data/output/' # For output\n",
    "\n",
    "file_path_benchmark = 'HGVS_2020_big_benchmark.tsv'\n",
    "file_path_polyscores = 'HGVS_2020_big_polyphen_scores.tsv'\n",
    "file_path_siftscores = 'HGVS_2020_big_sift_scores.tsv'\n",
    "file_path_vepbaseline = 'HGVS_2020_big_VEP_baseline.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6000  -  ./data/raw\\HGVS_2020_big_benchmark.tsv\n",
      "1200  -  ./data/raw\\HGVS_2020_big_polyphen_scores.tsv\n",
      "1200  -  ./data/raw\\HGVS_2020_big_sift_scores.tsv\n",
      "1200  -  ./data/raw\\HGVS_2020_big_VEP_baseline.tsv\n"
     ]
    }
   ],
   "source": [
    "csv_files = glob.glob(os.path.join(dir_data_raw, \"*.tsv\"))\n",
    "# loop over the list of csv files\n",
    "for f in csv_files:\n",
    "      \n",
    "    # read the csv file\n",
    "    df = pd.read_csv(f, sep='\\t')\n",
    "    # print the location and filename\n",
    "    print(len(df['# ID'].unique()), ' - ', f, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Exercise 2 \n",
    "### https://www.ncbi.nlm.nih.gov/clinvar/?term=APOE%5Bgene%5D&redir=gene - \n",
    "### Count 94 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Exercise 3.1\n",
    "### - What is the meaning of the highlighted symbols in the HGVS ID?\n",
    "### https://varnomen.hgvs.org/recommendations/DNA/variant/substitution/\n",
    "### ”>” = type of change is a substitution = >\n",
    "\n",
    "## Exercise 3.2\n",
    "\n",
    "### NC_000003.12:g.12599717C>G\tPathogenic\n",
    "### REF: https://varnomen.hgvs.org/bg-material/simple/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Exercise 4\n",
    "### NC_000001.11:g.224424520A>C\tS/R\tagT/agG\n",
    "### Lowest SIFT (0 = deleterious) score and highest Polyphen (1 = deleterious) \n",
    "### \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def parse_args():\n",
    "    \"\"\"\n",
    "        Parses inputs from the commandline.\n",
    "        :return: inputs as a Namespace object\n",
    "    \"\"\"\n",
    "    parser = argparse.ArgumentParser(description='Generates baseline model predictions.')\n",
    "    # Arguments\n",
    "    parser.add_argument('vep', help='a path to the VEP input file')\n",
    "    parser.add_argument('blosum', help='a path to the BLOSUM62 input file')\n",
    "    parser.add_argument('-o', dest='out_path', help='a path to write the output .tsv file with baseline model scores. This arguments is required!', required=True)\n",
    "    return parser.parse_args()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_blosum(path):\n",
    "    \"\"\"\n",
    "        Reads BLOSUM62 matrix file and stores in a 2-dimensional dictionary.\n",
    "        :param path: a str with the BLOSUM62 substitution matrix file path\n",
    "        :return: a 2-dimensional dict with amino acid (AA) substitution scores\n",
    "    \"\"\"\n",
    "    aas = []\n",
    "    aa_scores = []\n",
    "    with open(path, \"rb\") as f:\n",
    "        for line in f:\n",
    "            # Convert bytes to str\n",
    "            line = line.decode('UTF-8')\n",
    "            # Skip headers\n",
    "            if line.startswith('#') or line.startswith('x'):\n",
    "                continue\n",
    "            # Store AAs and scores (matrix is symmetric)\n",
    "            else:\n",
    "                aas.append(line.strip('\\n').split()[0])\n",
    "                aa_scores.append(line.strip('\\n').split()[1:len(line)])\n",
    "    # Create a 2-dimensional dictionary with AAs as keys and empty dictionaries as values\n",
    "    blosum_dict = {}\n",
    "    for aa in aas:\n",
    "        blosum_dict[aa] = {}\n",
    "        \n",
    "    #########################\n",
    "    ### START CODING HERE ###\n",
    "    #########################\n",
    "    # You need to fill in the dictionaries in blosum_dict for each key AA (amino acid) with the corresponding\n",
    "    # substitution scores between that AA and any other AA.\n",
    "    # Use aas list to loop over all amino acids:\n",
    "    # for i in range(len(aas)):\n",
    "        # Now loop over each score position for the AA with position i which are stored in the i-th list of aa_scores.\n",
    "        # These are the BLOSUM62 scores between the AA with position i and the AA with position j:\n",
    "        # for j in range(...):\n",
    "            # Get the score between the AA with position i and the AA with position j from aa_scores:\n",
    "            # score = ...\n",
    "            # For the dictionary of the i-th AA key of blosum_dict, store the j-th AA as a key and the score as a value:\n",
    "            # blosum_dict... = ...\n",
    "    #########################\n",
    "    ###  END CODING HERE  ###\n",
    "    #########################\n",
    "\n",
    "    return blosum_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def parse_vep(path):\n",
    "    \"\"\"\n",
    "        Reads VEP file and parses HGVS IDs and corresponding AA reference-mutation pairs.\n",
    "        :param path: a str with the VEP input file path\n",
    "        :return: three lists with HGVS IDs, reference AAs, and corresponding mutation AAs, respectively\n",
    "    \"\"\"\n",
    "\n",
    "    hgvs_ids = []\n",
    "    ref_aas = []\n",
    "    mut_aas = []\n",
    "    with open(path, \"rb\") as f:\n",
    "        # Read lines\n",
    "        for line in f:\n",
    "            # Convert bytes to str\n",
    "            line = line.decode('UTF-8').strip('\\n')\n",
    "            # Skip header\n",
    "            if line.startswith('#'):\n",
    "                continue\n",
    "            else:\n",
    "                # Get HGVS ID from the first column and append to the list\n",
    "                hgvs_ids.append(line.split('\\t')[0])\n",
    "                # Get amino acid mutation which is in the second column in the file\n",
    "                vars = line.split('\\t')[1]\n",
    "                #########################\n",
    "                ### START CODING HERE ###\n",
    "                #########################\n",
    "                # You need to get reference and mutation AAs from vars separately and append to the respective lists:\n",
    "                # ref_aas needs to contain reference amino acids;\n",
    "                # mut_aas needs to contain mutated amino acids.\n",
    "                # Have a look at vars to see how AAs can be separated from the string and think\n",
    "                # which string method you could use.\n",
    "                # Append the retrieved reference and mutation amino acids to the respective lists\n",
    "\n",
    "                #########################\n",
    "                ###  END CODING HERE  ###\n",
    "                #########################\n",
    "    return hgvs_ids, ref_aas, mut_aas\n",
    "\n",
    "def run_baseline(hgvs_ids, ref_aas, mut_aas, blosum_dict):\n",
    "    \"\"\"\n",
    "        Computes substitution scores for a dataset of SNPs using BLOSUM62 matrix.\n",
    "        :param hgvs_ids: a list of HGVS IDs obtained from parse_vep()\n",
    "        :param ref_aa: a list of corresponding reference AAs from parse_vep()\n",
    "        :param mut_aa: a list of corresponding mutation AAs from parse_vep()\n",
    "        :param blosum: a 2-dimensional dict of BLOSUM62 substitution matrix from parse_blosum()\n",
    "        :return: a list of calculated substitution scores\n",
    "    \"\"\"\n",
    "\n",
    "    # A list to store substitution scores for a dataset of SNPs\n",
    "    scores = []\n",
    "\n",
    "    #########################\n",
    "    ### START CODING HERE ###\n",
    "    #########################\n",
    "    # You need to calculate the score for each SNP using a for-loop.\n",
    "    # We need to have access to each HGVS ID, reference AA, and mutation AA. These can be found in\n",
    "    # hgvs_ids, ref_aas, and mut_aas, respectively. Note, these lists have the same length.\n",
    "    # You can use the following loop:\n",
    "    # for i in range(len(hgvs_ids)):\n",
    "        # Get reference and mutation AAs from the corresponding lists\n",
    "        # ref_aa = ...\n",
    "        # mut_aa = ...\n",
    "\n",
    "        # Compute BLOSUM62 substitution score the reference AA and the mutation AA stored in blosum_dict\n",
    "        # score = ...\n",
    "\n",
    "        # Append the score to scores\n",
    "\n",
    "    #########################\n",
    "    ###  END CODING HERE  ###\n",
    "    #########################\n",
    "\n",
    "    return scores\n",
    "\n",
    "def write_data(hgvs_ids, scores, out_filepath):\n",
    "    \"\"\"\n",
    "        Writes baseline model output to a .tsv file.\n",
    "        :param hgvs_ids: a list of HGVS IDs obtained from parse_vep()\n",
    "        :param scores: a list of corresponding BLOSUM62 substitution scores from run_baseline()\n",
    "        :param out_filepath: a str with the output .tsv file path\n",
    "    \"\"\"\n",
    "\n",
    "    # Open the file to write baseline model results\n",
    "    with open(out_filepath, 'w') as f:\n",
    "        # First line contains headers (tab-delimited HGVS ID and Score)\n",
    "        f.write('# ID\\tScore\\n')\n",
    "        # for SNP (its respective HGVS ID) and the calculated score\n",
    "        for id, score in zip(hgvs_ids, scores):\n",
    "            # Write HGVS ID and the calculated score to the file with tab separation\n",
    "            f.write(id + '\\t' + score + '\\n')\n",
    "\n",
    "        f.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def main():\n",
    "\n",
    "    # Process arguments\n",
    "    args = parse_args()\n",
    "    vep_path = args.vep\n",
    "    blosum_path = args.blosum\n",
    "    out_filepath = args.out_path\n",
    "\n",
    "    out_dir, out_filename = os.path.split(out_filepath)\n",
    "    # Check if output filename contains .tsv extension\n",
    "    if '.tsv' not in out_filename:\n",
    "        sys.exit(r'ERROR: filename \"%s\" in the output file path argument should contain .tsv extension!' % out_filename)\n",
    "\n",
    "    # Check if output directory exists\n",
    "    if not os.path.exists(out_dir):\n",
    "        sys.exit(r'ERROR: output directory \"%s\" to store baseline model output does not exist! Follow instructions in'\n",
    "                 r'the manual!' % out_dir)\n",
    "\n",
    "    # Parse BLOSUM62 matrix into a 2-dimensional dictionary\n",
    "    blosum_dict = parse_blosum(blosum_path)\n",
    "    # Parse VEP input file into lists of HGVS IDs, reference AAs, and corresponding mutation AAs\n",
    "    hgvs_ids, ref_aas, mut_aas = parse_vep(vep_path)\n",
    "    # Run baseline model\n",
    "    scores = run_baseline(hgvs_ids, ref_aas, mut_aas, blosum_dict)\n",
    "    # Write to a file\n",
    "    write_data(hgvs_ids, scores, out_filepath)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ad2bdc8ecc057115af97d19610ffacc2b4e99fae6737bb82f5d7fb13d2f2c186"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
